client/src/commands/getTcb.ts

```diff
@@ -0,0 +1,120 @@
+ * @param didRequestTcb persistent event from which notifications of
+ *        requested/udpated TCB virtual documents will be fired, prompting
+ *        VSCode to update the content of the document.
+ *        Subsequent calls of `executeGetTemplateTcb` should use the same value
+ *        for this parameter.
+ */
+async function executeGetTemplateTcb(
+    ngClient: AngularLanguageClient,
+    didRequestTcb: vscode.EventEmitter<vscode.Uri>): Promise<void> {
+  const activeEditor = vscode.window.activeTextEditor;
+  const client = ngClient.languageClient;
+  if (activeEditor === undefined || client === null) {
+    return;
+  }
+
+  const tcb = await client.sendRequest(lspext.GetTcb, {
+    textDocument: client.code2ProtocolConverter.asTextDocumentIdentifier(activeEditor.document),
+    position: activeEditor.selection.active,
+  });
+  if (tcb === undefined) {
+    return;
+  }
+
+  // Allocate a provider of documents corresponding to the `ng` URI scheme,
+  // which we will use to provide a virtual document with the TCB contents.
+  const tcbProvider: vscode.TextDocumentContentProvider = {
+    onDidChange: didRequestTcb.event,
+    provideTextDocumentContent: () => tcb.content,
+  };
+  const disposeNgProvider = vscode.workspace.registerTextDocumentContentProvider('ng', tcbProvider);
+  // Since `tcb.fileName` is an arbitrary filename and may contain "special"
+  // characters like '#' or '$' that can be interpreted as meaningful in a URI
+  // scheme if parsed, we have to be careful to define the URI explicitly.
+  const tcbUri = vscode.Uri.parse('').with({
+    scheme: 'ng',
+    authority: 'getTcb',
+    path: tcb.fileName,
+  });
+
+  const document = await vscode.workspace.openTextDocument(tcbUri);
+  // Notify VSCode of a change to the TCB virtual document, prompting it to
+  // re-evaluate the document content. This is needed to bust VSCode's document
+  // cache if someone requests a TCB that we previously opened.
+  // https://code.visualstudio.com/api/extension-guides/virtual-documents#update-virtual-documents
+  didRequestTcb.fire(tcbUri);
+  const editor = await vscode.window.showTextDocument(document, {
+    viewColumn: (vscode.window.activeTextEditor?.viewColumn ?? 1) + 1,
+    preserveFocus: true,
+  });
+
+  const sm = new SourceMap(tcb.content);
+  const highlightSelections = tcb.selections.map(
+      ({start, length}) => new vscode.Range(sm.toPosition(start), sm.toPosition(start + length)));
+  editor.setDecorations(TCB_HIGHLIGHT_DECORATION, highlightSelections);
+
+  // We immediately dispose of the provider so we can create a fresh provider
+  // for the `ng` scheme next time `getTcb` is requested.
+  disposeNgProvider.dispose();
+}
+
+/**
+ * Command getTemplateTcb displays a typecheck block for the template a user has
+ * an active selection over, if any.
+ */
+export function getTemplateTcb(ngClient: AngularLanguageClient) {
+  const didRequestTcb = new vscode.EventEmitter<vscode.Uri>();
```

Comment:
The return value of `getTemplateTcb` (namely `executeTemplateTcb`) is registered as a callback on the `getTcb` command. When that command is invoked `executeTemplateTcb` is called and we instantiate a document provider, alongside content for the TCB document.

However, the problem at this point is that VSCode updates documents lazily, and will not attempt to recompute a document's content without a notification that it needs to do so. So if we have already requested a TCB, say `a.typecheck.ts`, and then change something in our template and execute the `getTcb` command again, even though we will have updated the document provider and its content, the VSCode engine will not know to pick up these changes. Hence we associate an event emitter with each fresh document provider; VSCode will subscribe to the emitter, and when a new request for `getTcb` is made we fire the event to instruct VSCode to update the document content.

We could listen to changes in the active document, but that is more work than I want to include in this PR; it is a straightforward addition, but I think for these commits we should just get the command working without extra niceties.

Actually I just thought of a better model; I don't think there is any reason we have to reinstantiate a fresh document provider for each request. Again, I'll follow up after some more digging. Sorry, I have been very busy with other commitments so my responses here may be slow.

Reply:
Okay, so the other option I was thinking is that we can have a single document provider and recompute the TCB closed over that provider (and filename, highlights). So it would look something like

```
function makeExecuteGetTcb(...) {
  let currentHighlights;
  let currentTcbFileName;
  let currentTcbContents;
  function update() {
      tcb = getTcb(fileName);
      currentHighlights = getHighlights(tcp);
      currentTcbFileName = tcb.fileName;
      currentTcbContents = tcb.content;
  }
  class TcbProvider extends vscode.TextDocumentProvider {
    onDidChange: new EventEmitter(),
    provideTextDocumentContent(fileName) {
      return currentTcbContents;
    }
  }
  const provider = new TcbProvider();
  vscode.register(provider);
  return function execute() {
    update();
    provider.onDidChange.fire(currentTcbFileName);
    const doc = vscode.open(currentTcbFileName);
    vscode.addDecorations(doc, currentHighlights);
  }
}
```

I'll let you decide which you like more; in either case it's not the most elegant because the filename that we have to compute is not known before we make the getTcb request.

Let me know if you have any more questions; I hope the last comment was able to clear up some confusion if any.

Reply:
this is my thought with a single document provider. the only thing for the command `angular.getTemplateTcb` to do is just open the document by Uri. The provider will listen to the HTML change and refresh the tcb document.

```ts
class TcbDocumentContentProvider implements vscode.TextDocumentContentProvider, vscode.Disposable {
  private openFileSet = new Set<string>();

  constructor(private client: lsp.LanguageClient) {
    vscode.workspace.onDidOpenTextDocument(doc => {
      if (doc.uri.scheme === 'ngTcb') {
        this.openFileSet.add(doc.uri.query);
      }
    });
    vscode.workspace.onDidChangeTextDocument(doc => {
      if (this.openFileSet.has(doc.document.uri.path)) {
        const uri = vscode.Uri.parse('ngTcb: ngcheck.ts').with({query: doc.document.uri.path});
        this.onDidChangeEmitter.fire(uri);
      }
    });
    vscode.workspace.onDidCloseTextDocument(doc => {
      if (doc.uri.scheme === 'ngTcb') {
        this.openFileSet.delete(doc.uri.query);
      }
    });
    vscode.workspace.registerTextDocumentContentProvider('ngTcb', this);
  }

  dispose() {
    this.onDidChangeEmitter.dispose();
  }

  onDidChangeEmitter = new vscode.EventEmitter<vscode.Uri>();
  onDidChange = this.onDidChangeEmitter.event;

  provideTextDocumentContent(uri: vscode.Uri, token: vscode.CancellationToken):
      vscode.ProviderResult<string> {
    return this.client
        .sendRequest<GetTcbResponse>('ngTcb', {
          textDocument: lsp.TextDocumentIdentifier.create(uri.query),
          position: 0,
        })
        .then(tcb => tcb?.content);
  }
}
```
```ts
this.context.subscriptions.push(new TcbDocumentContentProvider(this.client));
```
```ts
function makeExecuteGetTcb(){
   let uri = vscode.Uri.parse('ngTcb: ngcheck.ts').with({query: activeEditor.document.uri.path});
    let doc = await vscode.workspace.openTextDocument(uri);  // calls back into the provider
    await vscode.window.showTextDocument(doc, {preview: false});
}
```

Reply:
@ayazhafiz with your implementation in the comment above, does that mean we can only keep track of one tcb at a time? Can a provider manage multiple files at a time? I don't think so, but looking at @ivanwonder's suggestion it seems possible?

Reply:
No, you can keep track of as many TCBs as you want if you have a single provider. In the example I give it only keeps track of one, but yes you can certainly keep track of several. However, I think "keeping track" of previously opened TCB documents is unnecessary, since (1) anytime someone requests a TCB you almost certainly want to make a request to the language server to get fresh TCB contents (I don't think trying to implement any kind of cache is worth it), and (2) if someone closes a TCB and then wants to reopen it, VSCode will have already cached the document and will keep that document content until the provider fires an event marking that the document should be reloaded.

So in my opinion, the only "important" thing is making sure that the event emitter for changed TCBs is properly exposed. To this end I feel the ways to  implement the provider here are the options between
- Creating a fresh document provider each time "getTcb" is requested, passing the same event emitter for requested file names (namely the file name corresponding to the requested TCB). This the current implementation.
- Making one document provider and closing over it, as done in Ivan's example or mine in the comment above Ivan's.

IMO, keeping track of previous document contents is unnecessary as explained above.

Reply:
> (1) anytime someone requests a TCB you almost certainly want to make a request to the language server to get fresh TCB contents (I don't think trying to implement any kind of cache is worth it),

I think I just record the HTML files for the TCB content and refresh the TCB content when the HTML files are changed. The TCB is managed by the uri, one TCB will be opened for the same uri. I have not tested it, but I think vscode will handle the uri. For example, when the user tries to view the TCB of file `files:/app/app.component.html`, we only need to open uri `ngtcb:/app/app.component.html`. The provider will handle the uri and provide the TCB content by sending the request to the service. and when the HTML changed, I will tell the provider the `ngtcb:/app/app.component.html` needs to be refreshed.

> (2) if someone closes a TCB and then wants to reopen it, VSCode will have already cached the document and will keep that document content until the provider fires an event marking that the document should be reloaded.

In my example, I think this will send a request to the service, because the vscode doesn’t know if the content is fresh, but it is ok to send a request(the language service has a cache for it) or add a cache here.

If I missed some info or misunderstood you, sorry for that. Because I think we can’t avoid listening to the HTML files, maybe you do it every time when open the TCB content. Just like you said it’s an option to achieve the function.

Some situations we need to be considered in the future.

- file deleted or renamed. If we need to close the TCB content or not.
- Sync scroll. Maybe we need to send a request to the service and scroll the TCB.
- file persistence. Maybe it’s not so important. After reload the vscode, the vscode will restore the layout and the opened file except for the TCB content.

So perhaps managed by uri is easier to understand for me.